common/core/flow_engine.c

```diff
@@ -0,0 +1,265 @@
+ * PRIVATE TYPEDEFS
+ *****************************************************************************/
+typedef struct {
+  flow_step_t **buffer;
+  array_list_t *array_list_config;
+} engine_ctx_t;
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+static flow_step_t *engine_buffer_0[ENGINE_STACK_DEPTH] = {0};
+static array_list_t engine_list_0;
+
+static flow_step_t *engine_buffer_1[ENGINE_STACK_DEPTH] = {0};
+static array_list_t engine_list_1;
+
+static engine_ctx_t engine_ctx[ENGINE_NUM_BUFFERS] = {
+    {&engine_buffer_0[0], &engine_list_0},
+    {&engine_buffer_1[0], &engine_list_1}};
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTION PROTOTYPES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void engine_initialize(void) {
+  /* Initialize all queues support by the engine */
+  for (uint8_t indx = 0; indx < ENGINE_NUM_BUFFERS; indx++) {
+    (void)engine_reset_flow(indx);
+  }
+}
+
+bool engine_reset_flow(uint8_t engine_buffer) {
+  if (ENGINE_NUM_BUFFERS <= engine_buffer) {
+    return false;
+  }
+
+  /* Zero-ise the content of the queue and the queue struct */
+  memzero(engine_ctx[engine_buffer].buffer,
+          ENGINE_STACK_DEPTH * (sizeof(flow_step_t *)));
+  memzero(engine_ctx[engine_buffer].array_list_config, sizeof(array_list_t));
+
+  /* Initialize the array_list with the appropriate parameters */
+  array_list_initialize(engine_ctx[engine_buffer].array_list_config,
+                        engine_ctx[engine_buffer].buffer,
+                        sizeof(flow_step_t *),
+                        ENGINE_STACK_DEPTH * (sizeof(flow_step_t *)));
+
+  return true;
+}
+
+bool engine_add_next_flow_step(uint8_t engine_buffer,
+                               const flow_step_t *flow_step_ptr) {
+  bool result = false;
+
+  if (ENGINE_NUM_BUFFERS <= engine_buffer) {
+    return result;
+  }
+
+  /* We want to store the pointer to the flow step inputted from the argument
+   * and NOT the actual content, therefore we need to pass the double pointer to
+   * the API */
+  const flow_step_t **flow_step_dptr = &flow_step_ptr;
+  result = array_list_insert(engine_ctx[engine_buffer].array_list_config,
+                             flow_step_dptr);
+  return result;
+}
```

Comment:
Should check here if `ENGINE_STACK_DEPTH` has been reached. A buffer overflow could be overlooked currently.